synapse/storage/databases/main/deviceinbox.py

```diff
@@ -229,10 +229,6 @@ async def get_messages_for_device(
             limit=limit,
         )
 
-        if not user_id_device_id_to_messages:
```

Comment:
I am unsure removing this is globally valid, but otherwise `self._limited_sync_and_check(0, 0)` still triggers an increment of the stream id in the sync token, hence `self._limited_sync_and_check(3, 1)` is failing and a `to_device` message is lost.

Reply:
I think if we're given a limit of zero we should bail *before* we do the call to `_get_device_messages`.

The reason for this check is performance: if the query didn't return any updates then we know there are know updates up until `to_stream_id`. In that case `last_processed_stream_id` may be an older stream id; returning it wouldn't be wrong it would just cause us to query the same range again next time.